src/server/qtquick/wtextureproviderprovider.cpp

```diff
@@ -0,0 +1,93 @@
+{
+    Q_OBJECT
+public:
+    RenderToImageJob(WOutputRenderWindow *renderWindow, WTextureProviderProvider *provider, QPromise<QImage> promise)
+        : m_renderWindow(renderWindow)
+        , m_provider(provider)
+        , m_promise(std::move(promise))
+    { }
+
+    void run() override {
+        if (m_promise.isCanceled())
+            return;
+        m_promise.start();
+        WSGTextureProvider *textureProvider = m_provider->wTextureProvider();
+        if (textureProvider && textureProvider->texture() && textureProvider->texture()->rhiTexture()) {
+            // Perform rhi texture read back
+            auto texture = textureProvider->texture()->rhiTexture();
+            qCInfo(qLcTextureProvider) << "Perform rhi texture read back for texture" << texture;
+            QRhiReadbackResult *rbResult = new QRhiReadbackResult;
+            QRhiCommandBuffer *cb;
+            m_renderWindow->rhi()->beginOffscreenFrame(&cb);
+            auto ub = m_renderWindow->rhi()->nextResourceUpdateBatch();
+            cb->beginComputePass(ub);
+            QRhiReadbackDescription rbd(texture);
+            ub->readBackTexture(rbd, rbResult);
+            cb->endComputePass(ub);
+            auto frameOpResult = m_renderWindow->rhi()->endOffscreenFrame();
+            if (frameOpResult == QRhi::FrameOpSuccess) {
+                m_promise.addResult(QImage(reinterpret_cast<const uchar *>(rbResult->data.constData()),
+                                           rbResult->pixelSize.width(),
+                                           rbResult->pixelSize.height(),
+                                           QImage::Format_RGBA8888_Premultiplied,
+                                           cleanupRbResult,
+                                           rbResult)); // TODO: Automatically get data format.
+            } else {
+                m_promise.setException(std::make_exception_ptr(std::runtime_error("Offscreen frame operation failed.")));
+            }
+        } else {
+            m_promise.setException(std::make_exception_ptr(std::runtime_error("Texture provider is not valid.")));
+        }
+        m_promise.finish();
+        disconnect();
+        deleteLater();
+    }
+    WOutputRenderWindow *const m_renderWindow;
+    WTextureProviderProvider *const m_provider;
+    QPromise<QImage> m_promise;
+};
+
+QFuture<QImage> WTextureProviderProvider::renderToImage()
+{
+    QPromise<QImage> promise;
+    auto future = promise.future();
+
+    auto job = new RenderToImageJob(outputRenderWindow(), this, std::move(promise));
+    if (outputRenderWindow()->inRendering()) {
+        QObject::connect(outputRenderWindow(), &WOutputRenderWindow::renderEnd, job, &RenderToImageJob::run, Qt::QueuedConnection);
+    } else {
+        // FIXME What if a new render process start before this job is executed?
+        WThreadUtil threadUtil(QQuickWindowPrivate::get(outputRenderWindow())->context->thread());
+        threadUtil.exec(job, &RenderToImageJob::run);
```

Comment:
The `exec` will block current thread, this is not a good idea, please `new WThreadUtil` and use `run`.